pkg/controller/reconciler/reconciler.go

```diff
@@ -120,6 +137,41 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 	return ctrl.Result{}, nil
 }
 
+func (r *Reconciler) addFinalizer(ctx context.Context, res common.PlatformObject) error {
+	l := log.FromContext(ctx)
+
+	// no finalizer action present => no finalizer to be added/checked for
+	if len(r.Finalizer) == 0 {
+		return nil
+	}
+
+	finalizerAdded := controllerutil.AddFinalizer(res, componentFinalizer)
+	if !finalizerAdded {
+		// finalizer already present
+		return nil
+	}
+
+	l.Info("adding finalizer")
+	if err := r.Client.Update(ctx, res); err != nil {
+		return fmt.Errorf("failed to add finalizer %s to %s", componentFinalizer, res.GetName())
+	}
+
+	return nil
+}
+
+func (r *Reconciler) removeFinalizer(ctx context.Context, res common.PlatformObject) error {
+	l := log.FromContext(ctx)
+
+	l.Info("removing finalizer")
+	controllerutil.RemoveFinalizer(res, componentFinalizer)
```

Comment:
nit: `controllerutil.RemoveFinalizer` returns whatever the finalizer has been removed or not, so you can early return if no finalizers have been removed which would avoid calling `Update`